src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -438,29 +443,20 @@ function beginUpdateQueue(
     // Second condition ignores top-level unmount callbacks if they are not the
     // last update in the queue, since a subsequent update will cause a remount.
     if (update.callback && !(update.isTopLevelUnmount && update.next)) {
-      const callbackUpdate = cloneUpdate(update);
-      if (callbackList && callbackList.last) {
-        callbackList.last.next = callbackUpdate;
-        callbackList.last = callbackUpdate;
-      } else {
-        callbackList = {
-          first: callbackUpdate,
-          last: callbackUpdate,
-          hasForceUpdate: false,
-        };
-      }
+      callbackList = callbackList || [];
+      callbackList.push(update.callback);
       workInProgress.effectTag |= CallbackEffect;
     }
     update = update.next;
   }
 
-  if (!queue.first && !queue.hasForceUpdate) {
-    // Queue is now empty
+  queue.callbackList = callbackList;
+
+  if (!queue.first && !queue.callbackList && !queue.hasForceUpdate) {
```

Comment:
Minor nit: do you need to dereference `queue.callbackList` here? Can you use `callbackList` directly?